core/src/apps/cardano/sign_tx.py

```diff
@@ -197,10 +204,52 @@ def _validate_outputs(
                 "Each output must have an address field or address_parameters!"
             )
 
+        if output.token_bundle:
+            _validate_token_bundle(output.token_bundle)
+
     if total_amount > LOVELACE_MAX_SUPPLY:
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_token_bundle(token_bundle: List[CardanoTokenGroupType]) -> None:
+    assets_count = sum([len(token_group.token_amounts) for token_group in token_bundle])
+
+    # This limitation is a mitigation measure to prevent sending
+    # large (especially change) outputs containing many tokens that Trezor
+    # would not be able to spend given that
+    # currently the full Cardano transaction is held in-memory.
+    # Once Cardano-transaction signing is refactored to be streamed, this
+    # limit can be lifted
+    if assets_count > MAX_ASSETS_PER_OUTPUT:
```

Comment:
Since this is not easy to discover during regular testing, it might be worthwhile to clearly document it in a place read by sw wallet developers integrating with trezor, and perhaps explicitly mention it to them in slack.